homeassistant/components/hassio/__init__.py

```diff
@@ -1003,10 +1005,23 @@ async def _async_refresh(
         """Refresh data."""
         if not scheduled:
             # Force refreshing updates for non-scheduled updates
-            try:
-                await self.hassio.refresh_updates()
-            except HassioAPIError as err:
-                _LOGGER.warning("Error on Supervisor API: %s", err)
+            # It `raise_on_auth_failed` is set, it means this is
+            # the first refresh. We background the update task in
+            # that case.
+            if raise_on_auth_failed:
+                self.config_entry.async_create_background_task(
+                    self.hass, self._async_refresh_updates(), "hassio_refresh_updates"
+                )
```

Comment:
I'm not sure we even need to do the background refresh since it won't write state when its done. AFAICT the change is relying on the next update to come around and do it.

Reply:
I tried this change and everything still seems like it works
```python
        if not scheduled and not raise_on_auth_failed:
            # Force refreshing updates for non-scheduled updates
            # If `raise_on_auth_failed` is set, it means this is
            # the first refresh and we do not want to delay
            # startup or cause a timeout so we only refresh the
            # updates if this is not a scheduled refresh and
            # we are not doing the first refresh.
            await self._async_refresh_updates()
```

Reply:
> I'm not sure we even need to do the background refresh since it won't write state when its done

We don't write state on our end, but we do ask the Supervisor to update its information. The next time the coordinator updates all entities, all information will be current.

This is to avoid stale information after startup (as I understood from @ludeeus).

> AFAICT the change is relying on the next update to come around and do it.

Which would be dictated by the Supervisor I guess? That might be right...

Reply:
This refresh isn't called again (it isn't ran on scheduled calls). It only runs if the user manual refreshes the data (hit the update button in the UI).

The scheduled update for checking for updates, is done by the Supervisor in general, which already locks and throttles this specific request.

Reply:
> The next time the coordinator updates all entities, all information will be current.

Isn't it going to update again before it does that?  I guess it wouldn't if its `scheduled` .. but I guess we don't know if the background task is finished either than?

Reply:
The idea is: fire and forget, when the data is updated we get it at some point the coordinator is freshing. This only applies to startup, as in all other cases, the service call to refresh (done from the UI) awaits it to complet. (Shown by the toasts in the bottom left of the UI).

This call only instructs the SU to update its external version information, which on startup, doesn't matter much if it happens instantly or takes longer. It may even fail fully at this point (previous data exists in the Supervisor).

The supervisor updates it on a schedule otherwise, locks and throttles any other scheduled/manual call as well. (A user can't spam the check for updates button).

Your proposed code change, I had originally as well. Especially, since the Supervisor already has a scheduled update. The comment from Ludeeus was, that this call existed to flush out any stale data on HA startup by requesting a refresh. This just does that.

Reply:
> This is to avoid stale information after startup (as I understood from @ludeeus).

That, and its used in automations/scripts/UI feedback

Reply:
First refresh is called
https://github.com/home-assistant/core/blob/6519b24319c571580294ade48738123d35dc7b12/homeassistant/components/hassio/__init__.py#L634

It now runs in the background.. in the above timeout case it takes > 20s
https://github.com/home-assistant/core/pull/109809/files#diff-7b5b8c277e4e67ad9b5f8ac4062977ae8d0f7be7cd2798da332e4ab4459f1d05R1012

Startup continues on to setup platforms while the background task run
https://github.com/home-assistant/core/blob/6519b24319c571580294ade48738123d35dc7b12/homeassistant/components/hassio/__init__.py#L637

Entities get created with the stale data because the background task is still running.

A refresh is triggered which gets the data up to date
https://github.com/home-assistant/core/blob/6519b24319c571580294ade48738123d35dc7b12/homeassistant/components/hassio/entity.py#L62
https://github.com/home-assistant/core/blob/6519b24319c571580294ade48738123d35dc7b12/homeassistant/components/hassio/entity.py#L153

Is it possible the background task is still running when the refresh gets triggered from the entity that needs the stats data being added

Reply:
Maybe this is completely overkill ...
```diff
diff --git a/homeassistant/components/hassio/__init__.py b/homeassistant/components/hassio/__init__.py
index b207b63c630..09c2755baeb 100644
--- a/homeassistant/components/hassio/__init__.py
+++ b/homeassistant/components/hassio/__init__.py
@@ -775,6 +775,7 @@ class HassioDataUpdateCoordinator(DataUpdateCoordinator):  # pylint: disable=has
         self._container_updates: defaultdict[str, dict[str, set[str]]] = defaultdict(
             lambda: defaultdict(set)
         )
+        self._background_refresh_updated_task: asyncio.Task | None = None
 
     async def _async_update_data(self) -> dict[str, Any]:
         """Update data via library."""
@@ -1009,10 +1010,22 @@ class HassioDataUpdateCoordinator(DataUpdateCoordinator):  # pylint: disable=has
             # the first refresh. We background the update task in
             # that case.
             if raise_on_auth_failed:
+                self._background_refresh_updated_task = (
                     self.config_entry.async_create_background_task(
-                    self.hass, self._async_refresh_updates(), "hassio_refresh_updates"
+                        self.hass,
+                        self._async_refresh_updates(),
+                        "hassio_refresh_updates",
+                    )
+                )
+            elif (
+                self._background_refresh_updated_task
+                and not self._background_refresh_updated_task.done()
+            ):
+                _LOGGER.warning(
+                    "Background task for refreshing updates still running, skipping"
                 )
             else:
+                self._background_refresh_updated_task = None
                 await self._async_refresh_updates()
 
         await super()._async_refresh(
```

Reply:
Maybe await the background task if it's not the first refresh and there is a background task that is not done? After the await, set the task to None. If there's no background task await a new update call instead.

Reply:
Sorry, but looking at the response by Ludeeus above, I think it is safe to just remove this call completely with your earliest suggestion.

Let's go with that 👍

Reply:
Done in [bd9c49a](https://github.com/home-assistant/core/pull/109809/commits/bd9c49a83a145d5f899a2590ef94c161df011f5f) (best viewed via the full diff though).